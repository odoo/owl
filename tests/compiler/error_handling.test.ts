import { renderToString, snapshotEverything, TestContext } from "../helpers";

snapshotEverything();

describe("error handling", () => {
  test("invalid xml", () => {
    expect(() => renderToString("<div>")).toThrow("Invalid XML in template");
  });

  test("nice warning if no template with given name", () => {
    const context = new TestContext();
    expect(() => context.renderToString("invalidname")).toThrow("Missing template");
  });

  test("cannot add twice the same template", () => {
    const context = new TestContext();
    context.addTemplate("test", `<t></t>`);
    expect(() => context.addTemplate("test", "<div/>", { allowDuplicate: true })).not.toThrow(
      "already defined"
    );
    expect(() => context.addTemplate("test", "<div/>")).toThrow("already defined");
  });

  // test("addTemplates throw if parser error", () => {
  //   const context = new TestContext();
  //   expect(() => {
  //     context.addTemplates("<templates><abc>></templates>");
  //   }).toThrow("Invalid XML in template");
  // });

  test("nice error when t-on is evaluated with a missing event", () => {
    expect(() => renderToString(`<div t-on="somemethod"></div>`)).toThrow(
      "Missing event name with t-on directive"
    );
  });

  test("error when unknown directive", () => {
    expect(() => renderToString(`<div t-best-beer="rochefort 10">test</div>`)).toThrow(
      "Unknown QWeb directive: 't-best-beer'"
    );
  });
});
