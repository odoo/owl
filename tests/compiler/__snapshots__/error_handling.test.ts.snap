// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error handling cannot add twice the same template 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  
  let block1 = createBlock(\`<div/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    return block1();
  }
}"
`;
